packages/material-ui-lab/src/TreeNode/TreeNode.js

```diff
@@ -0,0 +1,325 @@
+
+const styles = {
+  /* Styles applied to the root element. */
+  root: {
+    listStyle: 'none',
+    margin: 0,
+    padding: 0,
+    outline: 0,
+    '&:focus > $content': {
+      outline: 'auto 1px',
+    },
+  },
+  /* Styles applied to the `role="group"` element. */
+  group: {
+    margin: 0,
+    padding: 0,
+    marginLeft: 26,
+  },
+  /* Styles applied to the tree node content. */
+  content: {
+    width: '100%',
+    display: 'flex',
+    alignItems: 'center',
+  },
+  /* Styles applied to the tree node icon and collapse/expand icon. */
+  iconContainer: {
+    marginRight: 2,
+    width: 24,
+    minWidth: 24,
+    display: 'flex',
+    justifyContent: 'center',
+  },
+  /* Styles applied to the label element. */
+  label: {},
+};
+
+const isPrintableCharacter = str => {
+  return str && str.length === 1 && str.match(/\S/);
+};
+
+const TreeNode = React.forwardRef(function TreeNode(props, ref) {
+  const {
+    classes,
+    className,
+    children,
+    collapseIcon,
+    expandIcon,
+    icon,
+    nodeId,
+    label,
+    onClick,
+    onFocus,
+    onKeyDown,
+    ...other
+  } = props;
+  const {
+    icons,
+    toggle,
+    isExpanded,
+    isFocusable,
+    focus,
+    focusNextNode,
+    focusPreviousNode,
+    focusFirstNode,
+    focusLastNode,
+    isFocused,
+    handleLeftArrow,
+    expandAllSiblings,
+    setFocusByFirstCharacter,
+    handleNodeMap,
+    handleFirstChars,
+  } = React.useContext(TreeViewContext);
+
+  const nodeRef = React.useRef(null);
+  const contentRef = React.useRef(null);
+  const handleRef = useForkRef(nodeRef, ref);
+
+  let stateIcon = null;
+
+  const expandable = Boolean(children);
+  const expanded = isExpanded ? isExpanded(nodeId) : false;
+  const focused = isFocused ? isFocused(nodeId) : false;
+  const focusable = isFocusable ? isFocusable(nodeId) : false;
+
+  if (icons.defaultExpandIcon && expandable && !expanded) {
+    stateIcon = icons.defaultExpandIcon;
+  }
+  if (icons.defaultCollapseIcon && expandable && expanded) {
+    stateIcon = icons.defaultCollapseIcon;
+  }
+
+  if (expandIcon && expandable && !expanded) {
+    stateIcon = expandIcon;
+  }
+
+  if (collapseIcon && expandable && expanded) {
+    stateIcon = collapseIcon;
+  }
+
+  const stateIconsProvided = icons && (icons.defaultCollapseIcon || icons.defaultExpandIcon);
+
+  let startAdornment = null;
+
+  if (expandable) {
+    if (icons.defaultParentIcon || icon) {
+      startAdornment = icon || icons.defaultParentIcon;
+    }
+  } else if (icons.defaultEndIcon || icon) {
+    startAdornment = icon || icons.defaultEndIcon;
+  }
+
+  const handleClick = event => {
+    // only process click events that directly happened on this tree item
+    if (!contentRef.current.contains(event.target)) {
+      return;
+    }
+
+    if (!focused) {
+      focus(nodeId);
+    }
+
+    if (expandable) {
+      toggle(nodeId);
+    }
+
+    if (onClick) {
+      onClick(event);
```

Comment:
Hello, I think it would be interesting to add the `nodeId` so we can actually do something when displaying TreeNode in a `.map` call.
```suggestion
      onClick(event, nodeId);
```

Reply:
@raphaelbadia I don’t mind adding this but what do you think you could do with this over `onNodeExpanded` ?

Reply:
my bad, didn't see this function. It's no use then !